### PR TITLE
Api 37289 hardcode wsdl

### PR DESCRIPTION
**Pull request by acovrig**
## Summary
- Use hard-coded WSDL namespaces instead of doing a lookup
- Gracefully fallback to doing a lookup when there isn't a hardcoded value

## Related issue(s)
- [API-37289](https://jira.devops.va.gov/browse/API-37289)

## Testing done
- [x] *New code is covered by unit tests*
- I ran `rspec` with a debugger in either side of the cache check, both sides are currently covered by existing tests.
- StatsD test removed as it doesn't always hit that now if the cache hits.

## What areas of the site does it impact?
Anything that hits BGS, so most of claims_api

## Acceptance criteria
- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb [hunk 1]
```diff
@@ -19,7 +19,7 @@
         # 2: connection_wsdl_get - duration of WSDL request cycle
         # 3: connection_post - how long does the post itself take for the request cycle
         # 4: parsed_response - how long to parse the response
-        expect(ClaimsApi::Logger).to receive(:log).exactly(4).times
+        expect(ClaimsApi::Logger).to receive(:log).exactly(3).times
```
Comment (nihil2501) [inline modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb:22]: I think we should maintain the test coverage in this spec file over the legacy code path until that code path is actually gone. Relatedly, this spec file can be more explicit about exercising the different stages for our rollout.  Those stages correspond to the combinations of the two booleans `defined?` and `enabled?`.

But I don't see any Flipper logic included. I believe @FonzMP wanted a quick toggle back to WSDL fetching during the rollout period.

File: modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb [hunk 2]
```diff
@@ -40,18 +40,6 @@
...
-      VCR.use_cassette('claims_api/bgs/claimant_web_service/find_poa_by_participant_id',
-                       allow_playback_repeats: true) do
-        allow_any_instance_of(BGS::OrgWebService).to receive(:find_poa_history_by_ptcpnt_id).and_return({})
-
-        %w[establish_ssl_connection connection_wsdl_get connection_post parsed_response].each do |event|
-          expect { subject.find_poa_by_participant_id('does-not-matter') }
-            .to trigger_statsd_measure("api.claims_api.local_bgs.#{event}.duration")
-        end
-      end
-    end
```
Comment (nihil2501) [inline modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb:54]: It doesn't seem horrible to assert about this component that it is emitting the correct `statsd`  entries. Why is this being removed?

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb [hunk 3]
```diff
@@ -158,6 +158,36 @@ module UpdatePoaRequest
...
+          )
+        module FindPersonBySSN
+          DEFINITION =
+            Action.new(
+              service: PersonWebService::DEFINITION,
+              name: 'findPersonBySSN',
+              key: 'PersonDTO'
+            )
+        end
+      end
```
Comment (nihil2501) [inline modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:189]: Are we wanting to skip WSDL fetching for this action as soon as this code lands?

**Review by nihil2501 — CHANGES_REQUESTED** [review]
Some requested changes enumerated in the comments, but a couple of things related to the overall rollout plan.

Are we no longer planning to include a Flipper check if the action definition check passes? As a reminder, I think we discussed a strategy of:
- opting into the fetch-less implementation on a per-action basis by deploying individual additions to action definitions
- an emergency precaution to switch back to WSDL fetching globally with a Flipper flag

In particular, the logic could happen in this order with code that boils down to `defined? && enabled?` so that the Flipper DB check is only conducted if necessary.

Also, in its current form, this PR would immediately switch to the fetch-less implementation for all currently defined actions, whereas I believe we wanted to start from not doing it for any of the actions that currently happen in product. If we remove (or comment out) the existing definitions that correspond to code paths that use `LocalBGS` in production, this will surface an annoying fact about the spec suite in its current form.

Basically `local_bgs_refactored_spec.rb` is a duplicate of `local_bgs_spec.rb` which has the purpose of testing `LocalBGSRefactored` in the exact same way that `LocalBGS` is currently tested. And a couple of the current action definitions were just added to make that spec pass.  To get to a starting point of no BGS actions that currently happen in production using the fetch-less implementation to start, we need to find a way to resolve this conflict. One way is to remove (or skip) `local_bgs_refactored_spec.rb` for now and then find some way to remember to bring it back once fetch-less is fully rolled out or at least the actions it uses are rolled out, or changing how that spec is written. That isn't the only way, but skipping it is probably the simplest.

The `VdcBean` definitions can be left in to make other tests pass since those aren't used in production yet and can be fetch-less from the start.

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb (hunk 3 above)
Comment (nihil2501) [reply modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:189]: Also a nit pick, but if we do want this, maybe add an empty line above line 181.

File: modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb (hunk 2 above)
Comment (FonzMP) [reply modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb:54]: I think we could achieve re-establishing this test if we just remove the `connection_wsdl_get`

File: modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb (hunk 1 above)
Comment (FonzMP) [reply modules/claims_api/spec/lib/claims_api/local_bgs_spec.rb:22]: yeah, I definitely want to put this on a flipper feature for a safe introduction / testing phase. On a sync now with @acovrig and discussing. - thanks for the heads up and good eye @nihil2501

File: modules/claims_api/lib/bgs_service/local_bgs_refactored.rb [hunk 4]
```diff
@@ -1,5 +1,8 @@
 # frozen_string_literal: true
 
+require 'claims_api/bgs_client'
+require 'claims_api/bgs_client/definitions'
+require 'claims_api/bgs_client/error'
```
[the remark this answers is not in the document]
Comment (FonzMP) [reply modules/claims_api/lib/bgs_service/local_bgs_refactored.rb:5]: I feel like we have a ton of requires - would love to resolve the issue with needing to add these. Assuming it's a config thing. I think it's worth investigating.

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb (hunk 3 above)
Comment (FonzMP) [reply modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:189]: https://github.com/department-of-veterans-affairs/vets-api/pull/17185/files#r1648017735 - Yeah, like this on a Flipper Feature.

File: modules/claims_api/lib/bgs_service/local_bgs_refactored.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (acovrig) [reply modules/claims_api/lib/bgs_service/local_bgs_refactored.rb:5]: I agree, but they were required when I ran it from `bundle exec rails c`. It may work w/out them from `rails s`?

[the remark this answers is not in the document]
Comment (nihil2501) [reply modules/claims_api/lib/bgs_service/local_bgs_refactored.rb:5]: Steps to reproduce? This is what I see on my machine in Rails console:
```
> bundle exec rails c -s
irb(main):001> require 'bgs_service/local_bgs_refactored'
=> true
irb(main):002> ClaimsApi::LocalBGSRefactored::BAD_GATEWAY_EXCEPTIONS
=> [ClaimsApi::BGSClient::Error::ConnectionFailed, ClaimsApi::BGSClient::Error::SSLError, ClaimsApi::BGSClient::Error::TimeoutError]
```

File: modules/claims_api/lib/bgs_service/local_bgs.rb [hunk 5]
```diff
@@ -287,12 +289,30 @@ def make_request(endpoint:, action:, body:, key: nil, namespaces: {}, transform_
...
+            wsdl = log_duration(event: 'connection_wsdl_get', endpoint:) do
+              connection.get("#{Settings.bgs.url}/#{endpoint}?WSDL")
+            end
+            namespace = Hash.from_xml(wsdl.body).dig('definitions', 'targetNamespace').to_s
+          end
+        else
+          wsdl = log_duration(event: 'connection_wsdl_get', endpoint:) do
+            connection.get("#{Settings.bgs.url}/#{endpoint}?WSDL")
+          end
+          namespace = Hash.from_xml(wsdl.body).dig('definitions', 'targetNamespace').to_s
```
Comment (nihil2501) [inline modules/claims_api/lib/bgs_service/local_bgs.rb:312]: This logic can be expressed as something like this in order to remove duplication and also for the optimization of only checking the DB for the Flipper toggle when it is relevant by placing the definition test before the Flipper test. Also notice the `error` log level in the case of an unexpected exception:
```ruby
begin
  namespace = 
    LocalBGSRefactored::FindDefiniton
      .for_service(endpoint)
      .bean.namespaces
      .target
rescue LocalBGSRefactored::FindDefiniton::NotDefinedError
  # Expected during rollout.
rescue => e
  ClaimsApi::Logger.log(
    'local_bgs',
    detail: "local BGS FindDefinition Error: #{e.message}",
    level: :error
  )
end

if namespace.blank? || !Flipper.enabled?(:claims_api_hardcode_wsdl)
  wsdl_response =
    log_duration(event: 'connection_wsdl_get', endpoint:) do
      connection.get("#{Settings.bgs.url}/#{endpoint}?WSDL")
    end

  namespace =
    Hash
      .from_xml(wsdl_response.body)
      .dig('definitions', 'targetNamespace')
      .to_s
end
```

This is a good candidate for extracting to its own function, where the expression can probably look even more minimal/readable with early returns, but it's not necessary.

**Review by nihil2501 — CHANGES_REQUESTED** [review]
Again, when this rolls out, all currently hardcoded service definitions will switch to the WSDL-less strategy as soon as the Flipper is toggled on. I believe @FonzMP was keen on switching over a minimal set to start. See this previous [review comment](https://github.com/department-of-veterans-affairs/vets-api/pull/17185#pullrequestreview-2131121047) for possible caveats you may experience if paring down the starting list.

File: modules/claims_api/lib/bgs_service/local_bgs_refactored.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (acovrig) [reply modules/claims_api/lib/bgs_service/local_bgs_refactored.rb:5]: Interesting, on my machine, just running `bundle exec rails c` throws a `uninitialized constant ClaimsApi::LocalBGSRefactored::FindDefinition::BGSClient (NameError)`  for `find_definition.rb:22`.

[the remark this answers is not in the document]
Comment (nihil2501) [reply modules/claims_api/lib/bgs_service/local_bgs_refactored.rb:5]: Oh, without evaluating any expression in the repl at all? That's so odd.

**Comment by acovrig** [comment]
> Again, when this rolls out, all currently hardcoded service definitions will switch to the WSDL-less strategy as soon as the Flipper is toggled on. I believe @FonzMP was keen on switching over a minimal set to start. See this previous [review comment](https://github.com/department-of-veterans-affairs/vets-api/pull/17185#pullrequestreview-2131121047) for possible caveats in the spec suite you may experience if paring down the starting list.

I'm not sure I'm following, are you suggesting I:
* add a `defined?` method to each definition so we can enable them in prod for each deploy? Wouldn't we just add/remove the definition itself (because of the rescue)?
* add a flipper for each definition so we can enable/disable them at will (instead of waiting for a deploy)?\
* something else I haven't connected the dots to yet

File: config/features.yml [hunk 6]
```diff
@@ -171,6 +171,10 @@ features:
     actor_type: user
     description: Diverts codepath to LocalBGSRefactored
     enable_in_development: true
+  claims_api_hardcode_wsdl:
```
Comment (FonzMP) [inline config/features.yml:174]: should we `lighthouse_` this ... there's quite a few "claims" namespaces in vets-api currently. May provide a bit more clarity it's ours and not someone else's.

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb [hunk 7]
```diff
@@ -86,7 +86,7 @@ module EBenefitsBenefitClaimStatusWebServiceBean
           Bean.new(
             path: 'EBenefitsBnftClaimStatusWebServiceBean',
             namespaces: Namespaces.new(
-              target: 'http://services.share.benefits.vba.va.gov/',
+              target: 'http://claimstatus.services.ebenefits.vba.va.gov/',
```
Comment (FonzMP) [inline modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:89]: interesting, why the change here? was the former not actually a namespace for this service?

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb [hunk 8]
```diff
@@ -158,6 +158,37 @@ module UpdatePoaRequest
...
+      ##
+      # PersonWebServiceBean
+      #
+      module PersonWebServiceBean
+        DEFINITION =
+          Bean.new(
+            path: 'PersonWebServiceBean',
+            namespaces: Namespaces.new(
+              target: 'http://person.services.vetsnet.vba.va.gov/',
+              data: 'http://person.services.vetsnet.vba.va.gov/'
```
Comment (FonzMP) [inline modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:170]: do we need the duplication? seems weird to have two different items here.. unless it's to prepend the `/data` part?

**Comment by nihil2501** [comment]
> > Again, when this rolls out, all currently hardcoded service definitions will switch to the WSDL-less strategy as soon as the Flipper is toggled on. I believe @FonzMP was keen on switching over a minimal set to start. See this previous [review comment](https://github.com/department-of-veterans-affairs/vets-api/pull/17185#pullrequestreview-2131121047) for possible caveats in the spec suite you may experience if paring down the starting list.
> 
> I'm not sure I'm following, are you suggesting I:
> 
> * add a `defined?` method to each definition so we can enable them in prod for each deploy? Wouldn't we just add/remove the definition itself (because of the rescue)?
> * add a flipper for each definition so we can enable/disable them at will (instead of waiting for a deploy)?\
> * something else I haven't connected the dots to yet

The first point. Removing definitions that we are not currently feeling bold enough to go WSDL-less as soon as Flipper is toggled on.

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb (hunk 8 above)
Comment (nihil2501) [reply modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:170]: Yeah if there isn't a distinct second namespace that gets used inside the SOAP body for this bean, then the one parent namespace will already apply to all elements.

File: modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb (hunk 7 above)
Comment (nihil2501) [reply modules/claims_api/app/clients/claims_api/bgs_client/definitions.rb:89]: Yeah, I had that entered in incorrectly originally.

File: modules/claims_api/lib/bgs_service/local_bgs.rb (hunk 5 above)
Comment (nihil2501) [reply modules/claims_api/lib/bgs_service/local_bgs.rb]: This piece looks good.

The main reason for the blocked PR, which is still pending, is that a bunch of calls are going to switch to WSDL-less as soon as the Flipper toggle is on unless we comment out those hardcoded definitions.

**Review by nihil2501 — CHANGES_REQUESTED** [review]
Looks like there are 2 small points left and @FonzMP's approving review that are left to address on this PR:
- Pass the connection to that namespace method as an argument to reuse the one that was instantiated in the calling function (make_request), rather than making a new instance by invoking a new connection method.
- Set the namespace data: nil for PersonWebServiceBean

**Review by nihil2501 — APPROVED** [review]
LGTM

**Review by stevenjcumming — APPROVED** [review]

**Review by nihil2501 — APPROVED** [review]
LGTM
